Внимание на комментарии, нужна доработка на бэке.

Daria-pugacheva opened the pull request:
На фронте авторизация и регистрация подвязаны (в т.ч. для тестирования, как работает, сделала так, что добавление инструментов доступно только для зарегистрированных. Потом эту опцию будем добавлять, где надо).

НО для того, чтобы работала регистрация, надо исправить ряд моментов на бэке. Я сделала соответствующие TODO в UserDto, UserConverter и UserController.
Описываю словами, где неприятности:
1. Странно срабатывает валидация UssrDto - несмотря на то, что количество символов адекватное, вылетает 400-я ошибка и сообщение, что символов должно быть не более 255 и т.п. Из-за этого невозможно произвести регистрацию пользователя.
2. Если валидацию убрать нафиг (я пробовала закомментировать для теста), то выскакивает другая неприятность - при конвертации UserDto в User не хватает информации о ролях пользователя. Эту информацию пользователь сам не вводит. Я предлагаю для всех, кто регистрируется, при преобразовании ДТО-шки в сущность по умолчанию присваивать роль user. И из списка ролей, как говорили, убрать дифференциацию арендатор/арендодатель. А пусть там будут только user и admin. С админом потом сообразим, как его регить. Т.е. здесь тоже надо доработать.
3. И еще один момент с ошибками: Когда возникают ошибки валидации UserDto, то на фронт летит не просто список текстовых ошибок, а список ошибок, в которых лежат объекты, в которых два поля: название поля и ошибка в нем. Я пока не знаю, как на фронте такой многоуровневый список выводить в alert. Предлагаю над этим подумать или в части захвата на фронте или в UserController ловить эти тексты ошибок при валидации и отправлять на фронт в виде одноуровневого списка со строковыми описаниями.
Думаю, мою эту ветку с правками на фронте можно влить в девелоп и оптимально, чтобы уже с имеющимся работающим по авторизации/регистрации фронтом, был доработан бэк по этим нескольким моментам.

Daria-pugacheva committed:
На фронте авторизация и регистрация подвязаны (в т.ч. для тестирования, как работает, сделала так, что добавление инструментов доступно только для зарегистрированных. Потом эту опцию будем добавлять, где надо).
НО для того, чтобы работала регистрация, надо исправить ряд моментов на бэке. Я сделала соответствующие TODO в UserDto, UserConverter и UserController.
Описываю словами, где неприятности:
1. Странно срабатывает валидация UssrDto - несмотря на то, что количество символов адекватное, вылетает 400-я ошибка и сообщение,что символов должно быть не более 255 и т.п. Из-за этого невозможно произвести регистрацию пользователя.
2. Если валидацию убрать нафиг (я пробовала закомментировать для теста), то выскакивает другая неприятность - при конвертации UserDto в User не хватает информации о ролях пользователя. Эту информацию пользоатель сам не вводит. Я предлагаю для всех, кто регистрируется, при преобразовании ДТО-шки в сущность по умолчанию присваивать роль user. И из списка ролей, как говорили, убрать дифференциацию арендатор/арендодатель. А пусть там будут только user и admin. С админом потом сообразим, как его регить. Т.е. здесь тоже надо доработать.
3. И еще один момент с ошибками: Когда возникают ошибки валидации UserDto, то на фронт летит не просто список текстовых ошибок, а список ошибок, в которых лежат объекты, в которых два поля: название поля и ошибка в нем. Я пока не знаю, как на фронте такой многоуровневый список выводить в alert. Предлагаю над этим подумать или в части захвата на фронте или в UserController ловить эти тексты ошибок при валидации и отправлять на фронт в виде одноуровневого списка со строковыми описаниями.
Думаю, мою эту ветку с правками на фронте можно влить в девелоп и оптимально, чтобы уже с имеющимся работающим по авторизации/регистрации фронтом, был доработан бэк по этим нескольким моментам.

Haazaad commented on source/core/src/main/java/ru/letmerent/core/dto/UserDto.java:
```diff
@@ -21,7 +21,7 @@
 @ToString
 @Builder
-public class UserDto {
+public class UserDto {//TODO: Разобраться с валидацией: он работает как-то криво (в телеграмме отправила фото)
```
Поправлено в https://github.com/siporqueno/let-me-rent/pull/31/files#diff-040648ce469a4f624cc2b995ede87a2036cbca70898ffa6ae51974e2ab8cbfcd

Можно исключить

Haazaad commented on source/core/src/main/java/ru/letmerent/core/converters/UserConverter.java:
```diff
@@ -36,7 +36,7 @@ public UserDto userToUserDtoConverter(User user) {
     }
 
-    public User userDtoToUserConverter(UserDto userDto) {
+    public User userDtoToUserConverter(UserDto userDto) { //TODO:разобраться, как поступаем с ролями, когда пользователь регистрируется: юзер при регистрации роль не вводит и вылетает ошибка при попытке сохранить польователя
```
Поправлено в https://github.com/siporqueno/let-me-rent/pull/31/files#diff-976870fbb725e5837b618c0429637879a081059d23437bb8ea4a104de06d0935

Можно исключить

Daria-pugacheva replied:
Я по коду что-то не поняла, что ты здесь поправил... Ты же просто условия местами поменял, не? А при этом все равно должна роль добавляться. И при этом не вижу, чтобы в случае отсутствия стала добавляться роль просто user. или у меня глаза дырявые?....

Haazaad replied:
Кто сказал, что должна добавляться роль?

Haazaad commented on source/core/src/main/java/ru/letmerent/core/controllers/UserController.java:
```diff
@@ -62,7 +62,7 @@ public ResponseEntity<?> registerNewUser(@Valid @RequestBody UserDto userDto) {
         userService.saveUser(userDto);
         return new ResponseEntity<>(HttpStatus.CREATED);
-    }
+    }//TODO: если возникают ошибки валидации, то на фронт летит сложный объект ApplicationError, в котором лежать Violation, а в них поля "ИМЯ поля" и "сообщение об ошибке". Сложно пока это отразить на фронте - или на бэке надо сделать лист строк,или на фронте научиться их отражать.
```
Зачем делать лист строк?
В ответ прилетает объект ApplicationError, на фронте надо понять, что в userMessage массив или строка, в помощь дока https://docs.angularjs.org/api/ng/function/angular.isArray

Daria-pugacheva replied:
ну, надо поковыряться, да с фронтом. Обозначила, как то, до чего сегодня не добралась и опыта нет, чтобы раскрутить быстро :) Вдруг кто слету знает, в т.ч. Татьяна, вроде, с фронтом успешно какие-то фишечки подкручивала

Haazaad replied:
Тогда какой отношение данное TODO имеет к контроллеру?
Не стоит нагромождать код

tatiananosova replied:
@Haazaad  Проблема на бэке, если заполнено только одно (любое) поле, возращаются  только ошибки валидации из UserController, не возвращаются ошибки пустых полей. И эта (UserDto) валидация не работает совсем, то есть можно создать пользователя без имейла.

Daria-pugacheva committed:
Merge remote-tracking branch 'origin/develop' into LMR-33

Daria-pugacheva committed:
По итогам правок Артема на бэке и моих дработок, все вчерашние проблемы решены.
Авторизация работает, регистрация тоже.
При ошибках ввода данных (нулевой логин, еmail, пароль) , а также при совпадении логина, почты с имеющимися в базе и расхождении пароля и подтверждения выскакивают alert-ы с информацией о том, что не так.
В UserDto я добавила описания ошибок в валидацию,чтобы быо наглядно на фронте. И в UserController поменяла текст ошибок на русский вариант.
Думаю, можно мержить в develop без проблем эту ветку.
Один баг был, на который я натыкалась при проверке: если была сначала ошибка, например, что "такой адрес эл.почты уже существует", а после нее я просто удаляла данные из поля, то вместо ошибки, что "Поле с эл.почтой должно быть заполнено" периодически продолжала выпадать про "такой адрес эл.почты уже существует". Эта проблема решалась путем Shift F5. Не знаю, можно ли как-то ее нивелировать, чтобы работало, как часы.

siporqueno reviewed (approved):
1) Надо бы ещё сделать чтобы кнопка авторизации справа вверху после авторизации исчезала и рядом появлялась кнопка выйти из профиля, которая соответственно исчезает после выхода и вновь появляется кнопка авторизоваться.

Daria-pugacheva commented:
По Комментарию Николая сделаю вечерком, перед мержем добавлю изменения в  ветку.

tatiananosova committed:
Доработала аллерты регистрации

tatiananosova commented on source/core/src/main/resources/static/registration/registration.js:
```diff
@@ -1,18 +1,27 @@
+                    for (let i = 0; i < response.data.userMessage.length; i++) {
+                        result += response.data.userMessage[i].message + "\r\n\r\n"
+                    }
+                    alert(result.toString());
```
Всегда, при возникновении ошибки, перед текстом первой ошибки пишется undefined, как убрать и откуда это не нашла.

Daria-pugacheva committed:
Доработано: если пользователь уже авторизован, то в строке навигации вместо ссылки Авторизация появляется кнопка "Выйти из профиля", по нажатию на которую происходит выход из текущего профиля юзера.
Аналогичная кнопка оставлена и на странице авторизации, когда человек только что залогинился. На случай, если сразу передумал. Сейчас реализовано так.
Как вариант, можно сделать, чтобы после авторизации перекидывало на главную страницу (тогда на странице авторизации кнопку "выйти из своего профиля" можно будет убрать). Это сделать легко. Татьяна, Николай, какой вариант вам больше нравится?

Daria-pugacheva committed:
Merge remote-tracking branch 'origin/LMR-33' into LMR-33

Daria-pugacheva committed:
Починила косяк с undefined при выводе сообщения об ошибках валидации.

Commit:
Фикс валидации

Daria-pugacheva committed:
Merge remote-tracking branch 'origin/bugfix/fix-auth-validation' into LMR-33

# Conflicts:
#	source/core/src/main/java/ru/letmerent/core/dto/UserDto.java

Daria-pugacheva committed:
В ветку добавлны правки Артема по валидации + я дописала к аннотациями message для корректного отображения на фронте
